### PR TITLE
Added JUNIT support

### PR DESCRIPTION
**Pull request by rahul2393**
[Issue](https://github.com/aquasecurity/trivy/issues/93)

### CODE REVIEW

**Comment by rahul2393** [comment]
@knqyf263 Please take a look into this, I need the format in which to output the result, I just put the basic Hudson format

**Comment by knqyf263** [comment]
@rahul2393 Thanks. As jeff-cook mentioned, is it difficult to use a template like GitLab?
https://github.com/aquasecurity/trivy/issues/93#issuecomment-614688270

**Comment by rahul2393** [comment]
No, It is not difficult, can you please share the exact format of the template here so that I can use it

**Comment by knqyf263** [comment]
I'm not so familiar with Junit XML, but your XML looks fine. We don't have to filter a result based on severities because a user can specify it themself with `--severity` option.

```
<?xml version="1.0" ?>
<testsuits>
  <testsuite name="alpine:3.10.2" tests="2" errors="0" failures="2" time="0.003">
    <testcase classname="openssl-1.1.1c-r0" name="CVE-2020-1967" time="0">
      <failure type="description" message="" >
       openssl: information disclosure in fork() 
      </failure>
      <failure type="severity" message="" >HIGH</failure>
    </testcase>
    <testcase classname="openssl-1.1.1c-r0" name="CVE-2019-1547" time="0">
     ...
    </testcase>
  <system-out><![CDATA[sdtout!]]></system-out>
  <system-err><![CDATA[stderr!]]></system-err>
</testsuite>
```

**Comment by rahul2393** [comment]
@knqyf263 please check now

**Comment by knqyf263** [comment]
Thank you for the quick response, but I meant we could use a template like [gitlab.tpl](https://github.com/aquasecurity/trivy/blob/master/contrib/gitlab.tpl). It would be great if we can use the template to generate the similar XML file you generated by Go.

**Comment by rahul2393** [comment]
Added junit.tpl in contrib folder, also added test
@knqyf263

**Comment by knqyf263** [comment]
@rahul2393 Perfect! Would you update README as well? I think we can add the example next to the following section.
https://github.com/aquasecurity/trivy#save-the-results-using-a-template

File: contrib/junit.tpl [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" ?>
+<testsuites>
+{{ range . }}
+	{{- $failures := 0 }}
+    {{- range .Vulnerabilities -}}
+		{{- $failures = sum $failures 2 -}}
+	{{- end -}}
+	<testsuite tests="1" failures="{{ $failures }}" time="" name="{{  .Target }}">
+		{{- if not (eq .Type "") }}
+			<properties>
```
Comment (knqyf263) [inline contrib/junit.tpl:10]:
```suggestion
             <properties>
```

File: contrib/junit.tpl [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" ?>
+<testsuites>
+{{ range . }}
+	{{- $failures := 0 }}
+    {{- range .Vulnerabilities -}}
+		{{- $failures = sum $failures 2 -}}
+	{{- end -}}
+	<testsuite tests="1" failures="{{ $failures }}" time="" name="{{  .Target }}">
```
Comment (knqyf263) [inline contrib/junit.tpl:8]: I'm not so familiar with JUnit, so I just want to ask you. Isn't it weird that tests=1 and failures=10? I'm guessing "tests" is always greater than "failures".

Comment (rahul2393) [reply contrib/junit.tpl:8]: Here in the template, I assumed that 1 test per image and each failure are reported against the one test ran on image

Comment (rahul2393) [reply contrib/junit.tpl:8]: @knqyf263 Updated the Readme and indentation, let me know if you want to update the tests format

Comment (knqyf263) [reply contrib/junit.tpl:8]: It sounds fine. Let's keep it as is.

File: contrib/junit.tpl [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" ?>
+<testsuites>
+{{ range . }}
+	{{- $failures := 0 }}
+    {{- range .Vulnerabilities -}}
+	{{- $failures = sum $failures 2 -}}
```
Comment (knqyf263) [inline contrib/junit.tpl:6]: I'm just curious why 2 is added. I scanned alpine:3.10.2 and detected 5 vulnerabilities, but the XML file said 10 failures.

File: contrib/junit.tpl [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" ?>
+<testsuites>
+{{ range . }}
+	{{- $failures := 0 }}
+    {{- range .Vulnerabilities -}}
```
Comment (knqyf263) [inline contrib/junit.tpl:5]:
```suggestion
	{{- range .Vulnerabilities -}}
```

It looks like this template mixes tab and space.

File: contrib/junit.tpl [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+	<testsuite tests="1" failures="{{ $failures }}" time="" name="{{  .Target }}">
+		{{- if not (eq .Type "") }}
+        <properties>
+            <property name="type" value="{{ .Type }}"></property>
+        </properties>
+        {{- end -}}
+        {{ range .Vulnerabilities }}
+        <testcase classname="{{ .PkgName }}-{{ .InstalledVersion }}" name="{{ .VulnerabilityID }}" time="">
+            <failure message={{ .Title | printf "%q" }} type="description">{{ .Description | printf "%q" }}</failure>
+            <failure message="" type="severity">{{ .Vulnerability.Severity }}</failure>
```
Comment (knqyf263) [inline contrib/junit.tpl:17]: After the test on CircleCI, I found CircleCi didn't show the second `failure`.

image

We probably should embed severity into `name` in `testcase` like 
```
"【{{ .Vulnerability.Severity }}】 {{ .VulnerabilityID }}"

Comment (rahul2393) [reply contrib/junit.tpl]: @knqyf263 done

**Comment by z0mbix** [comment]
Hi, when is the next tagged release due that will contain this feature?

**Comment by knqyf263** [comment]
@z0mbix We're planning to release the latest version this week.

**Comment by knqyf263** [comment]
I'm sorry for the delay, but I've released v0.9.2 just now.
https://github.com/aquasecurity/trivy/releases/tag/v0.9.2

**Pull request merged**